File: packages/ai-config/src/schema/runtime.schema.ts
import { z } from 'zod';
import { ProviderSchema } from './provider.schema.js';
import { ModelSchema } from './model.schema.js';

const ContextDefaultSchema = z.object({
  provider: z.string().optional(),
  model: z.string().optional(),
  fallback: z.string().optional(),
});

export const RuntimeConfigSchema = z.object({
  mode: z.enum(['merge', 'replace']).default('replace'),
  providers: z.array(ProviderSchema).default([]),
  models: z.array(ModelSchema).default([]),
  profiles: z.record(z.string(), z.record(z.string(), z.string())).default({}),
  allowlist: z.array(z.string()).default([]),
  defaults: z
    .object({
      provider: z.string().optional(),
      model: z.string().optional(),
      byContext: z.record(z.string(), ContextDefaultSchema).default({}),
    })
    .default({ byContext: {} }),
});

export type RuntimeConfig = z.infer<typeof RuntimeConfigSchema>;

File: packages/ai_config/schema/runtime_schema.py
from typing import Dict, List, Optional

from pydantic import BaseModel, Field

from .provider_schema import ProviderSchema
from .model_schema import ModelSchema


class ContextDefaultSchema(BaseModel):
    provider: Optional[str] = None
    model: Optional[str] = None
    fallback: Optional[str] = None


class DefaultsSchema(BaseModel):
    provider: Optional[str] = None
    model: Optional[str] = None
    by_context: Dict[str, ContextDefaultSchema] = Field(default_factory=dict)


class RuntimeConfigSchema(BaseModel):
    mode: str = "replace"
    providers: List[ProviderSchema] = Field(default_factory=list)
    models: List[ModelSchema] = Field(default_factory=list)
    profiles: Dict[str, Dict[str, str]] = Field(default_factory=dict)
    allowlist: List[str] = Field(default_factory=list)
    defaults: DefaultsSchema = Field(default_factory=DefaultsSchema)
